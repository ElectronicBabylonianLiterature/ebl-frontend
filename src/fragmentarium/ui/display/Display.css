.Transliteration {
  list-style: none;
  padding: 0;
}

.Transliteration__ControlLine {
  color: darkgrey
}

.Transliteration__bracket,
.Transliteration__flag,
.Transliteration__subIndex,
.Transliteration__Determinative,
.Transliteration__Determinative .Transliteration__Reading,
.Transliteration__Determinative .Transliteration__Number,
.Transliteration__Determinative .Transliteration__UnidentifiedSign,
.Transliteration__Determinative .Transliteration__CompoundGrapheme,
.Transliteration__Determinative .Transliteration__Joiner {
  font-style: normal;
  font-variant: normal;
}

.Transliteration__Determinative + .Transliteration__Determinative > sup:before {
  content: ".";
}

.Transliteration__Reading,
.Transliteration__Number,
.Transliteration__UnidentifiedSign,
.Transliteration__CompoundGrapheme,
.Transliteration__Joiner,
.Transliteration__Divider,
.Transliteration__DocumentOrientedGloss,
.Transliteration__PhoneticGloss,
.Transliteration__LinguisticGloss {
  font-style: italic;
  font-variant: normal;
}

.Transliteration__Logogram {
  font-style: normal;
  font-variant: small-caps;
}
