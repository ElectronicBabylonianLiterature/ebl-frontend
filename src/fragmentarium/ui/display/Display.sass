
@mixin normal-font
  font-style: normal
  font-variant: normal

@mixin sumerian
  &--SUMERIAN
    @include normal-font
    letter-spacing: 0.1em

@mixin emesal
  &--EMESAL
    @include normal-font
    color: black
    letter-spacing: 0.1em

.Transliteration
  list-style: none
  padding: 0

  &__ControlLine
    color: darkgrey

  &__DollarAndAtLine,
  &__DollarAndAtLineWithParenthesis
    color: darkgrey
    text-indent: 1em

  &__ruling
      margin-bottom: 0.2em
      border-top: thin solid black

  &__Reading,
  &__Number,
  &__Divider
    @include normal-font

  &__Reading,
  &__UnidentifiedSign,
  &__Joiner,
    font-style: italic
    font-variant: normal

    .Transliteration__Determinative &,
    .Transliteration__Word--SUMERIAN &,
    .Transliteration__Word--EMESAL &
      @include normal-font

  &__bracket,
  &__flag,
  &__subIndex,
  &__modifier
    @include normal-font

  &__glossJoiner
    display: none

  &__Determinative
    @include normal-font

    & + & .Transliteration__glossJoiner
      display: inline

  &__wordSeparator,
  &__Word,
  &__Divider
    @include emesal
    @include sumerian

  &__Logogram,
  &__CompoundGrapheme,
  &__Grapheme
    font-style: normal
    font-variant: small-caps

  &__Reading,
  &__Number,
  &__Logogram,
  &__Divider,
  &__Grapheme,
  &__CompoundGrapheme,
  &__UnidentifiedSign,
  &__UnclearSign,
  &__Joiner,
  &__glossJoiner
    &--BROKEN_AWAY
      color: darkgrey

  &__BrokenAway,
  &__PerhapsBrokenAway,
  &__AccidentalOmission,
  &__IntentionalOmission,
  &__Removal
    @include normal-font

  &__Tabulation
    display: inline-block
    width: 1em

    & + .Transliteration__wordSeparator
      display: none
