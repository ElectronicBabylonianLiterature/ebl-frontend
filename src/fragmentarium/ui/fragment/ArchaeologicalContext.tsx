import _ from 'lodash'
import React from 'react'
import { Form, Col } from 'react-bootstrap'

function ArchaeologicalContext({ archaeology }: { archaeology? }): JSX.Element {
  return (
    <Form>
      <Form.Row>
        <Form.Group as={Col} controlId={_.uniqueId('Archaeology-')}>
          <Form.Label>Excavation number</Form.Label>
          <Form.Control
            plaintext
            readOnly
            value={archaeology?.excavationNumber || 'N/A'}
          />
        </Form.Group>
      </Form.Row>
    </Form>
  )
}

export default ArchaeologicalContext
