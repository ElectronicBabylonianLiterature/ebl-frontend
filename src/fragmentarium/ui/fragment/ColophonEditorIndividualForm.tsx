import React, { useState } from 'react'
import { Accordion, Button, Form, Row } from 'react-bootstrap'
import {
  Colophon,
  IndividualAttestation,
  IndividualType,
} from 'fragmentarium/domain/Colophon'
import _ from 'lodash'
import ListForm from 'common/List'
import { BrokenAndUncertainSwitches } from 'common/BrokenAndUncertain'
import FragmentService from 'fragmentarium/application/FragmentService'
import { getSelectField } from './ColophonEditorIndividualInputs'
import Bluebird from 'bluebird'

export interface IndividualProps {
  fragmentService: FragmentService
  individual: IndividualAttestation
  index: number
  onChange: (individual: IndividualAttestation, index: number) => void
}

export const ColophonIndividualsInput = ({
  fragmentService,
  onChange,
  label = 'Individuals',
  collapsed = false,
  individualsProp = [],
}: {
  fragmentService: FragmentService
  onChange: (field: keyof Colophon, value) => void
  label?: string
  collapsed?: boolean
  individualsProp?: ReadonlyArray<IndividualAttestation>
}): JSX.Element => {
  const [individuals, setIndividuals] = useState([...individualsProp])
  const getFormFields = (individual, _onChange, index) => (
    <Accordion key={index}>
      <Accordion.Toggle as={Button} variant="link" eventKey={index.toString()}>
        {`Individual ${index + 1}. ${individual.toString()}`}
      </Accordion.Toggle>
      <Accordion.Collapse eventKey={index.toString()}>
        <IndividualForm
          index={index}
          fragmentService={fragmentService}
          onChange={(_individual: IndividualAttestation, index: number) => {
            const _individuals = [...individuals]
            _individuals[index] = _individual
            setIndividuals(_individuals)
            onChange('individuals', _individuals)
          }}
          individual={individual}
          key={index}
        />
      </Accordion.Collapse>
    </Accordion>
  )
  return (
    <ListForm
      value={individuals}
      onChange={(_individuals: IndividualAttestation[]) => {
        setIndividuals(_individuals)
        onChange('individuals', _individuals)
      }}
      label={label}
      noun="Individual"
      defaultValue={new IndividualAttestation({})}
      collapsed={collapsed}
    >
      {getFormFields}
    </ListForm>
  )
}

type IndividualFieldName = 'type' | 'name' | 'sonOf' | 'grandsonOf' | 'family'

const setBrokenOrUncertain = (
  checked: boolean,
  field: 'isBroken' | 'isUncertain',
  keyField: 'name' | 'family' | 'sonOf' | 'grandsonOf' | 'type',
  individual: IndividualAttestation
): IndividualAttestation => {
  if (keyField === 'type') {
    return individual.setTypeField({
      ...individual[keyField],
      [field]: checked,
    })
  } else {
    return individual.setNameField(keyField, {
      ...individual[keyField],
      [field]: checked,
    })
  }
}

const getIndividualField = ({
  individualProps,
  key,
}: {
  individualProps: IndividualProps
  key: string
}): JSX.Element => {
  const { individual, fragmentService, onChange, index } = individualProps
  const fieldProps = getValueAndOptionsByKey(
    key as IndividualFieldName,
    individual,
    fragmentService
  )
  const props = {
    onChange,
    key,
    isClearable: true,
    placeholder: _.startCase(key),
    ...fieldProps,
  }

  const getBrokenOrUncertainMethod = (field: 'isBroken' | 'isUncertain') => (
    checked: boolean
  ) => {
    const _individual = setBrokenOrUncertain(
      checked,
      field,
      key as 'name' | 'family' | 'sonOf' | 'grandsonOf' | 'type',
      individual
    )
    onChange(_individual, index)
  }

  return (
    <Form.Group key={`${key}-col`}>
      <Form.Label>{_.startCase(key)}</Form.Label>
      {getSelectField({ key, individualProps, props })}
      <Row key={`${key}-row`}>
        <BrokenAndUncertainSwitches
          key={`${key}-broken-uncertain`}
          name={key}
          id={index}
          setBroken={getBrokenOrUncertainMethod('isBroken')}
          setUncertain={getBrokenOrUncertainMethod('isUncertain')}
          isBroken={individual[key]?.isBroken}
          isUncertain={individual[key]?.isUncertain}
        />
      </Row>
    </Form.Group>
  )
}

const IndividualForm = (individualProps: IndividualProps): JSX.Element => {
  const individualFieldNames = [
    'name',
    'sonOf',
    'grandsonOf',
    'family',
    'nativeOf',
    'type',
  ]
  return (
    <>
      {individualFieldNames.map((key) =>
        getIndividualField({ individualProps, key })
      )}
    </>
  )
}

const getValueAndOptionsByKey = (
  key: IndividualFieldName,
  individual: IndividualAttestation,
  fragmentService: FragmentService
): {
  value?: { value: string; label: string }
  options?: readonly { value: string; label: string }[]
  loadOptions?
  handleChange?
} => {
  return {
    ...(individual[key]?.value && {
      value: {
        value: individual[key]?.value as string,
        label: individual[key]?.value as string,
      },
    }),
    ...(key === 'type' && {
      options: Object.values(IndividualType).map((type) => ({
        value: type,
        label: type,
      })),
    }),
    ...(key !== 'type' && {
      loadOptions: getLoadOptionsMethod(fragmentService),
    }),
  }
}

export const getLoadOptionsMethod = (fragmentService: FragmentService) => (
  inputValue: string,
  callback: (options: { value: string; label: string }[]) => void
): Bluebird<void> =>
  fragmentService.fetchColophonNames(inputValue).then((entries) => {
    const options = entries.map((value) => ({
      value,
      label: value,
    }))
    callback(options)
  })
