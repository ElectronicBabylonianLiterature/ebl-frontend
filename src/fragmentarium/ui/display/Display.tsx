import React, { FunctionComponent } from 'react'
import _ from 'lodash'
import TransliterationHeader from 'fragmentarium/ui/view/TransliterationHeader'
import SessionContext from 'auth/SessionContext'
import Session from 'auth/Session'
import { Fragment } from 'fragmentarium/domain/fragment'
import { Glossary } from './Glossary'

import './Display.css'
import {
  Line,
  Token,
  Variant,
  NamedSign,
  Grapheme,
  Divider,
  UnknownSign
} from 'fragmentarium/domain/text'
import classNames from 'classnames'

function DefaultToken({
  token: { parts, value }
}: {
  token: Token
}): JSX.Element {
  return (
    <>
      {parts
        ? parts.map((token, index) => (
            <DisplayToken key={index} token={token} />
          ))
        : value}
    </>
  )
}

function UnknownSignComponent({ token }: { token: Token }): JSX.Element {
  const sign = token as UnknownSign
  const signs = {
    UnclearSign: 'x',
    UnidentifiedSign: 'X'
  }
  return (
    <>
      {signs[sign.type]}
      <sup className="Display__flag">{sign.flags.join('')}</sup>
    </>
  )
}

function NamedSignComponent({ token }: { token: Token }): JSX.Element {
  const namedSign = token as NamedSign
  return (
    <>
      {namedSign.name}
      {namedSign.subIndex !== 1 && (
        <sub className="Display__subIndex">{namedSign.subIndex || 'x'}</sub>
      )}
      <span className="Display__modifier">{namedSign.modifiers.join('')}</span>
      <sup className="Display__flag">{namedSign.flags.join('')}</sup>
      {namedSign.sign && (
        <>
          <span className="Display__bracket">(</span>
          <DisplayToken token={namedSign.sign} />
          <span className="Display__bracket">)</span>
        </>
      )}
      {namedSign.surrogate && !_.isEmpty(namedSign.surrogate) && (
        <>
          <span className="Display__bracket">&lt;(</span>
          {namedSign.surrogate.map((token, index) => (
            <DisplayToken key={index} token={token} />
          ))}
          <span className="Display__bracket">)&gt;</span>
        </>
      )}
    </>
  )
}

const tokens: ReadonlyMap<
  string,
  FunctionComponent<{ token: Token }>
> = new Map([
  ['UnknownNumberOfSigns', (): JSX.Element => <>â€¦</>],
  [
    'Variant',
    ({ token }: { token: Token }): JSX.Element => (
      <>
        {(token as Variant).tokens.map((token, index) => (
          <>
            {index > 0 ? '/' : null}
            <DisplayToken key={index} token={token} />
          </>
        ))}
      </>
    )
  ],
  ['Reading', NamedSignComponent],
  ['Logogram', NamedSignComponent],
  ['Number', NamedSignComponent],
  [
    'Divider',
    ({ token }: { token: Token }): JSX.Element => {
      const divider = token as Divider
      return (
        <>
          {divider.divider}
          <span className="Display__modifier">
            {divider.modifiers.join('')}
          </span>
          <sup className="Display__flag">{divider.flags.join('')}</sup>
        </>
      )
    }
  ],
  [
    'Grapheme',
    ({ token }: { token: Token }): JSX.Element => {
      const grapheme = token as Grapheme
      return (
        <>
          {grapheme.name}
          <span className="Display__modifier">
            {grapheme.modifiers.join('')}
          </span>
          <sup className="Display__flag">{grapheme.flags.join('')}</sup>
        </>
      )
    }
  ],
  ['UnclearSign', UnknownSignComponent],
  ['UnidentifiedSign', UnknownSignComponent]
])

function DisplayToken({ token }: { token: Token }): JSX.Element {
  const TokenComponent = tokens.get(token.type) || DefaultToken
  return (
    <span
      className={classNames([
        'Display__token',
        `Display__token--${token.type}`
      ])}
    >
      <TokenComponent token={token} />
    </span>
  )
}

function DisplayLine({
  line: { type, prefix, content },
  container = 'div'
}: {
  line: Line
  container?: string
}): JSX.Element {
  return React.createElement(
    container,
    { className: classNames(['Display__line', `Display__line--${type}`]) },
    <>
      <span>{prefix}</span>
      {content.map((token, index) => (
        <>
          {' '}
          <DisplayToken key={index} token={token} />
        </>
      ))}
    </>
  )
}

interface Props {
  fragment: Fragment
}

function Display({ fragment }: Props): JSX.Element {
  return (
    <>
      <TransliterationHeader fragment={fragment} />
      <ol className="Display__lines">
        {fragment.text.lines.map((line: Line, index: number) => (
          <li key={index}>
            <DisplayLine line={line} />
          </li>
        ))}
      </ol>
      <SessionContext.Consumer>
        {(session: Session): React.ReactNode =>
          session.hasBetaAccess() && <Glossary fragment={fragment} />
        }
      </SessionContext.Consumer>
    </>
  )
}

export default Display
