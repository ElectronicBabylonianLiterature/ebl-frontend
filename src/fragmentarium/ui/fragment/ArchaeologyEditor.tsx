import React, { ChangeEvent, Component, FormEvent } from 'react'
import _ from 'lodash'
import { Form, Col, Button } from 'react-bootstrap'
import Select, { ValueType } from 'react-select'
import {
  Archaeology,
  ArchaeologyDto,
  Findspot,
  SiteKey,
  excavationSites,
} from 'fragmentarium/domain/archaeology'
import { Fragment } from 'fragmentarium/domain/fragment'
import withData from 'http/withData'
import { FindspotService } from 'fragmentarium/application/FindspotService'

interface Props {
  archaeology?: Archaeology
  updateArchaeology: (archaeology: ArchaeologyDto) => Promise<Fragment>
  findspots: readonly Findspot[]
  disabled?: boolean
}

interface State {
  excavationNumber: string
  site: SiteKey
  isRegularExcavation: boolean
  findspotId: number | null
  findspot: Findspot | null
  error: Error | null
}

const excavationOptions = [
  {
    value: '',
    label: '-',
  },
  ..._.values(excavationSites).map((site) => ({
    value: site.name,
    label: site.name,
  })),
]

class ArchaeologyEditor extends Component<Props, State> {
  private isDirty = false
  private originalState: State
  private updateArchaeology: (archaeology: ArchaeologyDto) => Promise<Fragment>
  private findspots: ReadonlyMap<number, Findspot>
  private findspotOptions

  constructor(props: Props) {
    super(props)
    const archaeology = props.archaeology || {}

    this.state = {
      excavationNumber: archaeology.excavationNumber || '',
      site: (archaeology.site?.name || '') as SiteKey,
      isRegularExcavation: archaeology.isRegularExcavation ?? true,
      error: null,
      findspotId: archaeology.findspotId || null,
      findspot: archaeology.findspot || null,
    }
    this.originalState = { ...this.state }
    this.updateArchaeology = props.updateArchaeology

    this.findspots = new Map(
      props.findspots.map((findspot) => [findspot.id, findspot])
    )
    this.findspotOptions = props.findspots.map((findspot) => ({
      value: findspot.id,
      label: findspot.toString(),
    }))
  }

  updateState = (property: string) => (
    value: string | boolean | number | Findspot | null
  ): void => {
    const updatedState = {
      ...this.state,
      [property]: value,
    }
    this.isDirty = !_.isEqual(this.originalState, updatedState)
    this.setState(updatedState)
  }
  updateFindspotState = (
    findspotId: number | null,
    findspot: Findspot | null
  ): void => {
    this.updateState('findspotId')(findspotId)
    this.updateState('findspot')(findspot)
  }

  updateExcavationNumber = (event: ChangeEvent<HTMLInputElement>): void =>
    this.updateState('excavationNumber')(event.target.value)

  updateSite = (
    event: ValueType<typeof excavationOptions[number], false>
  ): void => this.updateState('site')(event?.value || '')

  updateIsRegularExcavation = (event: ChangeEvent<HTMLInputElement>): void =>
    this.updateState('isRegularExcavation')(event.target.checked)

  updateFindspot = (
    event: ValueType<typeof this.findspotOptions[number], false>
  ): void => {
    if (_.isNull(event)) {
      this.updateFindspotState(null, null)
    } else {
      this.updateFindspotState(
        event.value,
        this.findspots.get(event.value) || null
      )
    }
  }

  submit = (event: FormEvent<HTMLElement>): void => {
    event.preventDefault()

    this.updateArchaeology({
      ..._.omitBy(
        { ...this.state, error: null },
        (value) => _.isNil(value) || value === ''
      ),
    })
      .then((_updatedFragment) => {
        this.originalState = { ...this.state }
        this.isDirty = false
        this.setState(this.originalState)
      })
      .catch((error) =>
        this.setState({
          ...this.state,
          error: error,
        })
      )
  }

  renderExcavationNumberForm = (): JSX.Element => (
    <Form.Group as={Col} controlId={_.uniqueId('excavationNumber-')}>
      <Form.Label>Excavation number</Form.Label>
      <Form.Control
        type="text"
        value={this.state.excavationNumber}
        onChange={this.updateExcavationNumber}
      />
    </Form.Group>
  )
  renderExcavationSiteForm = (): JSX.Element => (
    <Form.Group as={Col} controlId={_.uniqueId('excavationSite-')}>
      <Form.Label>Excavation site</Form.Label>
      <Select
        aria-label="select-site"
        options={excavationOptions}
        value={{
          value: this.state.site,
          label: this.state.site,
        }}
        onChange={this.updateSite}
        isSearchable={true}
        isClearable
      />
    </Form.Group>
  )
  renderIsRegularExcavationForm = (): JSX.Element => (
    <Form.Group>
      <Form.Check
        type="checkbox"
        id={_.uniqueId('isRegularExcavation-')}
        label="Regular Excavation"
        checked={this.state.isRegularExcavation}
        onChange={this.updateIsRegularExcavation}
      />
    </Form.Group>
  )
  renderFindspotForm = (): JSX.Element => (
    <Form.Group as={Col} controlId={_.uniqueId('findspot-')}>
      <Form.Label>Findspot</Form.Label>
      <Select
        aria-label="select-findspot"
        options={this.findspotOptions}
        value={{
          value: this.state.findspotId,
          label: this.state.findspot?.toString(),
        }}
        onChange={this.updateFindspot}
        isSearchable={true}
        isClearable
      />
    </Form.Group>
  )

  render(): JSX.Element {
    return (
      <Form onSubmit={this.submit} data-testid="archaeology-form">
        <Form.Row>{this.renderExcavationNumberForm()}</Form.Row>
        <Form.Row>{this.renderExcavationSiteForm()}</Form.Row>
        <Form.Row>{this.renderIsRegularExcavationForm()}</Form.Row>
        <Form.Row>{this.renderFindspotForm()}</Form.Row>
        <Button
          variant="primary"
          type="submit"
          disabled={this.props.disabled || !this.isDirty}
        >
          Save
        </Button>
      </Form>
    )
  }
}

export default withData<
  {
    archaeology?: Archaeology
    updateArchaeology: (archaeology: ArchaeologyDto) => Promise<Fragment>
    disabled?: boolean
  },
  { findspotService: FindspotService },
  readonly Findspot[]
>(
  ({ archaeology, updateArchaeology, disabled, data: findspots }) => {
    return (
      <ArchaeologyEditor
        archaeology={archaeology}
        updateArchaeology={updateArchaeology}
        findspots={findspots}
        disabled={disabled}
      />
    )
  },
  (props) => props.findspotService.fetchFindspots()
)
