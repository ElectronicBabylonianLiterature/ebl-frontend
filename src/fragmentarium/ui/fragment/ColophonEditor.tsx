import React, { useState } from 'react'
import { Form, Button, Row } from 'react-bootstrap'
import { Fragment } from 'fragmentarium/domain/fragment'
import {
  ColophonOwnershipInput,
  ColophonStatusInput,
  ColophonTypeInput,
  ColophonNotesToScribalProcessInput,
  ProvenanceAttestationInput,
} from './ColophonEditorInputs'
import FragmentService from 'fragmentarium/application/FragmentService'
import { ColophonIndividualsInput } from './ColophonEditorIndividualForm'
import { Colophon } from 'fragmentarium/domain/Colophon'

interface Props {
  fragment: Fragment
  updateColophon: (colophon: Colophon) => Promise<void>
  disabled?: boolean
  fragmentService: FragmentService
}

const ColophonEditor: React.FC<Props> = ({
  fragment,
  disabled,
  updateColophon,
  fragmentService,
}) => {
  // ToDo:
  // 3. Ensure that the form loads and dumps data correctly
  // 4. Side panel
  // 5. Tests

  const { colophon } = fragment
  const [formData, setFormData] = useState<Colophon>({
    colophonStatus: colophon?.colophonStatus,
    colophonOwnership: colophon?.colophonOwnership,
    colophonTypes: colophon?.colophonTypes,
    originalFrom: colophon?.originalFrom,
    writtenIn: colophon?.writtenIn,
    notesToScribalProcess: colophon?.notesToScribalProcess,
    individuals: colophon?.individuals,
  })
  const [error, setError] = useState<Error | null>(null)

  const handleChange = (field: keyof Colophon, value) => {
    setFormData((prevState) => ({
      ...prevState,
      [field]: value,
    }))
  }

  const submit = async (): Promise<void> => {
    try {
      await updateColophon(formData)
    } catch (error) {
      setError(error as Error)
    }
  }

  return (
    <>
      <Form>
        <Row>
          <ColophonStatusInput
            colophonStatus={formData.colophonStatus}
            onChange={handleChange}
          />
        </Row>
        <Row>
          <ColophonTypeInput
            colophonTypes={formData.colophonTypes}
            onChange={handleChange}
          />
        </Row>
        <Row>
          <ColophonOwnershipInput
            colophonOwnership={formData.colophonOwnership}
            onChange={handleChange}
          />
        </Row>
        <Row>
          <ProvenanceAttestationInput
            {...{
              onChange: handleChange,
              fragmentService,
              fieldName: 'originalFrom',
              colophon: formData,
            }}
          />
        </Row>
        <Row>
          <ProvenanceAttestationInput
            {...{
              onChange: handleChange,
              fragmentService,
              fieldName: 'writtenIn',
              colophon: formData,
            }}
          />
        </Row>
        <ColophonIndividualsInput
          individualsProp={formData.individuals}
          onChange={handleChange}
          fragmentService={fragmentService}
        />
        <Row>
          <ColophonNotesToScribalProcessInput
            notesToScribalProcess={formData.notesToScribalProcess}
            onChange={handleChange}
          />
        </Row>
      </Form>
      <Button
        variant="primary"
        type="submit"
        aria-label="save-colophon"
        disabled={disabled || error !== null}
        onClick={(event) => {
          event.preventDefault()
          submit()
        }}
      >
        Save
      </Button>
    </>
  )
}

export default ColophonEditor
