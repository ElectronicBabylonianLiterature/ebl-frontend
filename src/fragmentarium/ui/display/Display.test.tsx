import React from 'react'
import { act, render, RenderResult } from '@testing-library/react'
import { Fragment } from 'fragmentarium/domain/fragment'
import complexText from 'test-support/complexTestText'
import WordService from 'dictionary/application/WordService'
import Display from './Display'
import { wordDto } from 'test-support/test-word'
import { MemoryRouter } from 'react-router-dom'
import { fragmentFactory } from 'test-support/fragment-fixtures'

jest.mock('dictionary/application/WordService')

let wordService
let fragment: Fragment
let element: RenderResult
let container: Element

beforeEach(async () => {
  wordService = new (WordService as jest.Mock<WordService>)()
  jest
    .spyOn(wordService, 'find')
    .mockImplementation(() => Promise.resolve(wordDto))
  fragment = fragmentFactory.build(
    {
      notes: 'lorem ipsum quia dolor sit amet',
      publication: 'Guod cigipli epibif odepuwu.',
      description:
        'Balbodduh lifuseb wuuk nasu hulwajo ho hiskuk riwa eldat ivu jandara nosrina abike befukiz ravsus.\nZut uzzejum ub mil ika roppar zewize ipifac vut eci avimez cewmikjov kiwso zamli jecja now.',
    },
    { associations: { text: complexText } }
  )
  await act(async () => {
    element = render(
      <MemoryRouter>
        <Display fragment={fragment} wordService={wordService} activeLine="" />
      </MemoryRouter>
    )
  })
  container = element.container
})

test(`Renders header`, () => {
  expect(container).toHaveTextContent(fragment.publication)
})

test('Snapshot', () => {
  expect(container).toMatchSnapshot()
})
