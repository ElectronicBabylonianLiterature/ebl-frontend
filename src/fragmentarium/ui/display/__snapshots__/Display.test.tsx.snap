// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
<div>
  <header>
    <p
      class="TransliterationHeader__description"
    >
      Balbodduh lifuseb wuuk nasu hulwajo ho hiskuk riwa eldat ivu jandara nosrina abike befukiz ravsus.
    </p>
    <p
      class="TransliterationHeader__description"
    >
      Zut uzzejum ub mil ika roppar zewize ipifac vut eci avimez cewmikjov kiwso zamli jecja now.
    </p>
    <p
      class="TransliterationHeader__publication"
    >
      (Publication: 
      Guod cigipli epibif odepuwu.
      )
    </p>
  </header>
  <ol
    class="Transliteration"
  >
    <li
      class="Transliteration__TextLine"
    >
      <span>
        1.
      </span>
       
      <span
        class="Transliteration__Word"
      >
        <span
          class="Transliteration__Reading"
        >
          kur
          <span
            class="Transliteration__modifier"
          >
            @v
          </span>
          <sup
            class="Transliteration__flag"
          >
            #!
          </sup>
        </span>
        <span
          class="Transliteration__Joiner"
        >
          .
        </span>
        <span
          class="Transliteration__Reading"
        >
          kur
          <sub
            class="Transliteration__subIndex"
          >
            x
          </sub>
          <span
            class="Transliteration__modifier"
          />
          <sup
            class="Transliteration__flag"
          />
          <span
            class="Transliteration__bracket"
          >
            (
          </span>
          <span
            class="Transliteration__Grapheme"
          >
            KUR
            <span
              class="Transliteration__modifier"
            />
            <sup
              class="Transliteration__flag"
            />
          </span>
          <span
            class="Transliteration__bracket"
          >
            )
          </span>
        </span>
        <span
          class="Transliteration__Joiner"
        >
          -
        </span>
        <span
          class="Transliteration__Logogram"
        >
          KUR
          <sub
            class="Transliteration__subIndex"
          >
            2
          </sub>
          <span
            class="Transliteration__modifier"
          >
            @v
          </span>
          <sup
            class="Transliteration__flag"
          >
            #!
          </sup>
        </span>
      </span>
       
      <span
        class="Transliteration__Word"
      >
        <span
          class="Transliteration__CompoundGrapheme"
        >
          |KUR.KUR|
        </span>
      </span>
       
      <span
        class="Transliteration__UnknownNumberOfSigns"
      >
        â€¦
      </span>
       
      <span
        class="Transliteration__UnidentifiedSign"
      >
        X
        <sup
          class="Transliteration__flag"
        />
      </span>
       
      <span
        class="Transliteration__UnclearSign"
      >
        x
        <sup
          class="Transliteration__flag"
        />
      </span>
       
      <span
        class="Transliteration__Variant"
      >
        <span
          class="Transliteration__Reading"
        >
          kur
          <span
            class="Transliteration__modifier"
          />
          <sup
            class="Transliteration__flag"
          />
        </span>
        /
        <span
          class="Transliteration__Logogram"
        >
          KUR
          <span
            class="Transliteration__modifier"
          />
          <sup
            class="Transliteration__flag"
          />
          <span
            class="Transliteration__bracket"
          >
            &lt;(
          </span>
          <span
            class="Transliteration__Reading"
          >
            kur
            <span
              class="Transliteration__modifier"
            />
            <sup
              class="Transliteration__flag"
            />
          </span>
          <span
            class="Transliteration__bracket"
          >
            )&gt;
          </span>
        </span>
      </span>
       
      <span
        class="Transliteration__Word"
      >
        <span
          class="Transliteration__Determinative"
        >
          <sup>
            <span
              class="Transliteration__Reading"
            >
              d
              <span
                class="Transliteration__modifier"
              />
              <sup
                class="Transliteration__flag"
              />
            </span>
            <span
              class="Transliteration__Joiner"
            >
              -
            </span>
            <span
              class="Transliteration__Reading"
            >
              kur
              <sub
                class="Transliteration__subIndex"
              >
                2
              </sub>
              <span
                class="Transliteration__modifier"
              />
              <sup
                class="Transliteration__flag"
              >
                ?
              </sup>
            </span>
          </sup>
        </span>
        <span
          class="Transliteration__Determinative"
        >
          <sup>
            <span
              class="Transliteration__Reading"
            >
              d
              <span
                class="Transliteration__modifier"
              />
              <sup
                class="Transliteration__flag"
              >
                !
              </sup>
            </span>
          </sup>
        </span>
        <span
          class="Transliteration__Reading"
        >
          kur
          <span
            class="Transliteration__modifier"
          />
          <sup
            class="Transliteration__flag"
          />
        </span>
        <span
          class="Transliteration__LinguisticGloss"
        >
          <sup>
            <span
              class="Transliteration__Reading"
            >
              kur
              <span
                class="Transliteration__modifier"
              />
              <sup
                class="Transliteration__flag"
              />
            </span>
          </sup>
        </span>
      </span>
       
      <span
        class="Transliteration__Word"
      >
        <span
          class="Transliteration__Reading"
        >
          kur
          <span
            class="Transliteration__modifier"
          />
          <sup
            class="Transliteration__flag"
          />
        </span>
        <span
          class="Transliteration__Joiner"
        >
          -
        </span>
        <span
          class="Transliteration__LinguisticGloss"
        >
          <sup>
            <span
              class="Transliteration__Reading"
            >
              kur
              <span
                class="Transliteration__modifier"
              />
              <sup
                class="Transliteration__flag"
              />
            </span>
            <span
              class="Transliteration__Joiner"
            >
              -
            </span>
            <span
              class="Transliteration__Reading"
            >
              kur
              <span
                class="Transliteration__modifier"
              />
              <sup
                class="Transliteration__flag"
              />
            </span>
          </sup>
        </span>
        <span
          class="Transliteration__Reading"
        >
          kur
          <span
            class="Transliteration__modifier"
          />
          <sup
            class="Transliteration__flag"
          />
        </span>
      </span>
    </li>
  </ol>
</div>
`;
